### PR TITLE
Color Matrices

### PR DESCRIPTION
**Pull request by bebon901**
Improved the code that fits the color matrices, overall reduced Delta E for all camera modules.

Only for review - do no merge. @davidplowman @naushir @njhollinghurst

### CODE REVIEW

File: utils/raspberrypi/ctt/colors.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# SIMPLE CODE TO CONVERT RGB VALUES TO LAB
+def RGB_to_LAB(RGB):  # where RGB is a 1x3 array.   e.g RGB = [100, 255, 230]
+    num = 0
+    XYZ = [0, 0, 0]
+    # converted all the three R, G, B to X, Y, Z
+    X = RGB[0] * 0.4124 + RGB[1] * 0.3576 + RGB[2] * 0.1805
+    Y = RGB[0] * 0.2126 + RGB[1] * 0.7152 + RGB[2] * 0.0722
+    Z = RGB[0] * 0.0193 + RGB[1] * 0.1192 + RGB[2] * 0.9505
+    # rounded off the values upto 4 decimal digit
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/colors.py:9]: Was this where we removed the code that was rounding the numbers? Maybe remove the comment if it isn't relevant any more.

File: utils/raspberrypi/ctt/colors.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    XYZ[0] = X / 255 * 100
+    XYZ[1] = Y / 255 * 100
+    XYZ[2] = Z / 255 * 100
+    XYZ[0] = float(XYZ[0]) / 95.047  # ref_X =  95.047   Observer= 2°, Illuminant= D65
+    XYZ[1] = float(XYZ[1]) / 100.0  # ref_Y = 100.000
+    XYZ[2] = float(XYZ[2]) / 108.883  # ref_Z = 108.883
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/colors.py:16]: Just wondering if it's necessary to cast to float here?

File: utils/raspberrypi/ctt/colors.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+def RGB_to_LAB(RGB):  # where RGB is a 1x3 array.   e.g RGB = [100, 255, 230]
+    num = 0
+    XYZ = [0, 0, 0]
+    # converted all the three R, G, B to X, Y, Z
+    X = RGB[0] * 0.4124 + RGB[1] * 0.3576 + RGB[2] * 0.1805
+    Y = RGB[0] * 0.2126 + RGB[1] * 0.7152 + RGB[2] * 0.0722
+    Z = RGB[0] * 0.0193 + RGB[1] * 0.1192 + RGB[2] * 0.9505
+    # rounded off the values upto 4 decimal digit
+
+    XYZ[0] = X / 255 * 100
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/colors.py:11]: Maybe a comment that the XYZ numbers need to be in the range 0 - 100 would be helpful?

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 4]
```diff
@@ -87,48 +117,128 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
+        b2 = ccm[7]
+        '''
+        COLOR MATRIX LOOKS AS BELOW
+        R1 R2 R3   Rval     Outr
+        G1 G2 G3  *  Gval  =  G
+        B1 B2 B3   Bval     B
+        Will be optimising 6 elements and working out the third element using 1-r1-r2 = r3
+        '''
+
+        x0 = [r1, r2, g1, g2, b1, b2]
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:169]: It looks like the next line is setting `x0` again, so maybe delete this one?

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 5]
```diff
@@ -87,48 +117,128 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
+        x0 = [2, -1, -0.4, 1.7, 0, -0.9]
+        result = minimize(guess, x0, args=(r, g, b, m_lab), tol=0.0000000001)  # This produces a color matrix which has the lowest delta E possible, based off the input data. Note it is impossible for this to reach zero since inperfect data.
+
+        Cam.log += ("\n \n Optimised Matrix Below: \n \n")
+        [r1, r2, g1, g2, b1, b2] = result.x
+        occm = [r1, r2, (1 - r1 - r2), g1, g2, (1 - g1 - g2), b1, b2, (1 - b1 - b2)]  # This is the optimised Color Matrix
+        Cam.log += str(occm)
+        Cam.log += str(ccm)
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:177]: Would the log be easier to follow if it just added that this is the original colour matrix?

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 6]
```diff
@@ -156,6 +266,50 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
+    g2 = ccm[4]
+    b1 = ccm[6]
+    b2 = ccm[7]
+    realrgb = []
+    for i in range(len(r)):
+        ir = r[i]
+        ig = g[i]
+        ib = b[i]
+        realrgb.append(colors.RGB_to_LAB([ir / 256 * r1 + ig / 256 * r2 + ib / 256 * (1 - r1 - r2), ir / 256 * g1 + ig / 256 * g2 + ib / 256 * (1 - g1 - g2), ir / 256 * b1 + ig / 256 * b2 + ib / 256 * (1 - b1 - b2)]))
+    # now compare that with m_lab and return numeric result, averaged for each patch
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:289]: Again, maybe the same kind of tidying here?

File: utils/raspberrypi/ctt/ctt_visualise.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+100x100 will be new color, 100x100 old color
+if 50px gap between pixels, image will be 7 gaps + 6 patches wide, and 5 gaps + 4 patches high.
+  y --->
+x
+|
+|
+'''
+
+
+def visualise_macbeth_chart(mrgb, orgb, nrgb, filename):
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_visualise.py:15]: Perhaps making the variable names a little less cryptic would help the next person!

File: src/ipa/rpi/vc4/data/imx296.json [hunk 8]
```diff
@@ -157,11 +157,6 @@
                     {
                         "shutter": [ 100, 5000, 10000, 20000, 30000 ],
                         "gain": [ 1.0, 2.0, 4.0, 8.0, 16.0 ]
-                    },
-                    "long":
-                    {
-                        "shutter": [ 1000, 30000, 60000, 90000, 120000 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 12.0 ]
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/imx296.json:164]: This seems to have been removed.

File: src/ipa/rpi/vc4/data/imx477.json [hunk 9]
```diff
@@ -155,18 +149,13 @@
...
-                        "shutter": [ 100, 5000, 10000, 20000, 33333 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 8.0 ]
-                    },
-                    "long":
-                    {
-                        "shutter": [ 100, 10000, 30000, 60000, 120000 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 12.0 ]
+                        "shutter": [ 100, 5000, 10000, 20000, 120000 ],
+                        "gain": [ 1.0, 2.0, 4.0, 6.0, 6.0 ]
                     }
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/imx477.json:159]: These have gone missing.

File: src/ipa/rpi/vc4/data/imx708_wide.json [hunk 10]
```diff
@@ -155,11 +155,6 @@
                     {
                         "shutter": [ 100, 5000, 10000, 20000, 120000 ],
                         "gain": [ 1.0, 2.0, 4.0, 6.0, 6.0 ]
-                    },
-                    "long":
-                    {
-                        "shutter": [ 1000, 30000, 60000, 90000, 120000 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 12.0 ]
                     }
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/imx708_wide.json:158]: Missing.

File: src/ipa/rpi/vc4/data/ov5647.json [hunk 11]
```diff
@@ -150,18 +149,13 @@
...
-                        "shutter": [ 100, 5000, 10000, 20000, 33333 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 8.0 ]
-                    },
-                    "long":
-                    {
-                        "shutter": [ 100, 10000, 30000, 60000, 120000 ],
-                        "gain": [ 1.0, 2.0, 4.0, 6.0, 12.0 ]
+                        "shutter": [ 100, 5000, 10000, 20000, 120000 ],
+                        "gain": [ 1.0, 2.0, 4.0, 6.0, 6.0 ]
                     }
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/ov5647.json:159]: Missing.

File: src/ipa/rpi/vc4/data/ov5647.json [hunk 12]
```diff
@@ -14,25 +14,25 @@
         {
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/ov5647.json:14]: Ditto about the wholesale changes.

File: src/ipa/rpi/vc4/data/imx708_wide.json [hunk 13]
```diff
@@ -408,6 +412,9 @@
         {
             "rpi.sharpen": { }
         },
+        {
+            "rpi.focus": { }
+        },
```
Comment (naushir) [inline src/ipa/rpi/vc4/data/imx708_wide.json:417]: The libcamera folks are likely to prefer a separate commit to introduce ``rpi.focus`` into all the tuning files.

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 14]
```diff
@@ -87,48 +117,128 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
     """
     ccm_tab = {}
     for Img in imgs:
-        Cam.log += '\nProcessing image: ' + Img.name
+        print('\nProcessing image: ' + Img.name)
```
Comment (naushir) [inline utils/raspberrypi/ctt/ctt_ccm.py:120]: Is this meant to change from Cam.log to print?

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 15]
```diff
@@ -87,48 +117,128 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
+            ir = r[p]
+            ig = g[p]
+            ib = b[p]
+            gargb.append(gamma([ir / 256 * r1 + ig / 256 * r2 + ib / 256 * (1 - r1 - r2), ir / 256 * g1 + ig / 256 * g2 + ib / 256 * (1 - g1 - g2), ir / 256 * b1 + ig / 256 * b2 + ib / 256 * (1 - b1 - b2)]))
+            galab.append(colors.RGB_to_LAB([ir / 256 * r1 + ig / 256 * r2 + ib / 256 * (1 - r1 - r2), ir / 256 * g1 + ig / 256 * g2 + ib / 256 * (1 - g1 - g2), ir / 256 * b1 + ig / 256 * b2 + ib / 256 * (1 - b1 - b2)]))
+        # Gamma After RGB / LAB
+        # We now want to spit out some data that shows how the optimisation has improved the color matricies
+        Cam.log += "Here are the Improvements"
+
+        # CALCULATE WORST CASE  / \e
```
Comment (naushir) [inline utils/raspberrypi/ctt/ctt_ccm.py:216]: deltae?

**Review by naushir — COMMENTED** [review]
One minor thing about the python changes.  It might be better to separate out the non-functional tidy-up changes from the functional changes.  This will be needed when we send the patches to the libcamera team.

**Comment by bebon901** [comment]
@davidplowman @naushir @njhollinghurst 
Updated files and resolved issues.

**Comment by bebon901** [comment]
Can you please also review the updated tuning files for cameras.
@davidplowman @naushir @njhollinghurst

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 16]
```diff
@@ -6,35 +6,74 @@
...
+    x = x / ((2 ** 8) - 1)  # takes 255 and scales it down to one
+    x = np.where(x < 0.04045, x / 12.92, ((x + 0.055) / 1.055) ** 2.4)
+    x = x * ((2 ** 16) - 1)  # takes one and scales up to 255
     return x
 
 
+def gamma(x):
+    # return (x *  * (1 / 2.4)  *  1.055 - 0.055)
+    e = []
+    for i in range(len(x)):
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:52]: Maybe `for colour in x:` might be a little more Pythonic. Or even
```
return [((colour / 255) ** (1 / 2.4) * 1.055 - 0.055) * 255 for colour in x]
```

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 17]
```diff
@@ -52,16 +91,22 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
-
     """
     convert reference colours from srgb to rgb
     """
-    m_srgb = degamma(m_rgb)
+    m_srgb = degamma(m_rgb)  # now in 16 bit color.
+
+    m_lab = []
+    for col in m_srgb:
+        m_lab.append(colors.RGB_to_LAB(col / 256))
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:101]: Could be another of those
```
m_lab = [colors.RGB_to_LAB(col / 256) for col in m_srgb]
```
?

File: utils/raspberrypi/ctt/ctt_ccm.py [hunk 18]
```diff
@@ -156,6 +292,50 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
...
+    # now compare that with m_lab and return numeric result, averaged for each patch
+    return (sumde(realrgb, m_lab) / 24)  # returns an average result of delta E
+
+
+def sumde(listA, listB):
+    global typenum, test_patches
+    sumde = 0
+    maxde = 0
+    patchde = []
+    for i in range(len(listA)):
```
Comment (davidplowman) [inline utils/raspberrypi/ctt/ctt_ccm.py:319]: Maybe `for listA_item, listB_item in zip(listA, listB):` would appeal more to Python fans?

**Comment by davidplowman** [comment]
Thanks, Ben. Latest version looking much closer now!

**Comment by bebon901** [comment]
Have updated - Hopefully those python fans out there enjoy my little changes... Even used the zip() function more than once!

**Comment by naushir** [comment]
This has now been merged upstream.

**Pull request closed without merging**